### PR TITLE
Use Astropy WCS for map pixel <-> world conversions and add unit support to Map.

### PR DESCRIPTION
**Pull request by Cadair**
This is an extension to #1352 and a fix for #912.

The current implementation of pixel to world coordinates and vice-a-versa in Map lead to rotation not being taken into account when the image is not aligned to the data axes. This will fix that problem.

Seeing how `rotate()` is now stable (I hope), this might be an issue more people encounter and in theory it is an easy ish fix.

The one thing this does is completely change the meaning of the `data_to_pixel` method, if anyone can shed some light on wtf the purpose of the original method was we can discuss keeping it!!

ping @drewleonard42 @ayshih @ehsteve @dpshelio

### CODE REVIEW

**Comment by dpshelio** [comment]
I believed the `data_to_pixel` is to obtain from a certain Solar-X,Y coordinate in a map which pixel corresponds.  Which is practically the same thing you've done using `wcs` - which I think is the proper way to do it. I think this is related with the SSW [`mk_map_xp`](http://hesperia.gsfc.nasa.gov/ssw/gen/idl/maps/mk_map_xp.pro) and [`mk_map_yp`](http://hesperia.gsfc.nasa.gov/ssw/gen/idl/maps/mk_map_yp.pro), or a similar one (which I cannot find now).

However, I wonder whether in `pixel_to_data` now we can try with non-real pixels (i.e., neagative ones or larger than the image size).

File: sunpy/map/mapbase.py [hunk 1]
```diff
@@ -458,41 +458,75 @@ def _validate(self):
...
+
+        y : `~astropy.units.Quantity`
+            Coordinate of the CTYPE2 axis. (Normally solar-y).
+        """
+
+        x, y = self.wcs.wcs_pix2world(np.array([[x, y]]), origin)[0]
+
+        # WCS always outputs degrees.
+        x *= u.deg
+        y *= u.deg
```
Comment (dpshelio) [inline sunpy/map/mapbase.py]: does wcs returns degrees without the `units`? is that something in `sunpy` or comes from `astropy`?

Comment (Cadair) [inline sunpy/map/mapbase.py]: yeah astropy.wcs.WCS doesn't return with units, but it is always in degrees.

**Comment by Cadair** [comment]
So this implementation will allow negative pixels etc, becuase astropy WCS does.

File: sunpy/map/mapbase.py (hunk 1 above)
Comment (dpshelio) [inline sunpy/map/mapbase.py]: Does astropy has all in units - as we try in sunpy? if so we may consider to ask them to provide it as quantities, don't you think?

**Comment by dpshelio** [comment]
Negative pixels is fine with me.

File: sunpy/map/mapbase.py (hunk 1 above)
Comment (astrofrog) [inline sunpy/map/mapbase.py]: For legacy reasons I don't know if we can change these values to be quantities now. But a new generalized WCS framework is in the works and we can definitely make sure that this returns quantities where appropriate.

File: sunpy/map/mapbase.py [hunk 2]
```diff
@@ -168,11 +168,11 @@ def __repr__(self):
     def wcs(self):
         w2 = astropy.wcs.WCS(naxis=2)
         w2.wcs.crpix = [self.reference_pixel['x'], self.reference_pixel['y']]
-        w2.wcs.cdelt = [self.scale['x'], self.scale['y']]
-        w2.wcs.crval = [self.reference_coordinate['x'], self.reference_coordinate['y']]
+        w2.wcs.cdelt = u.Quantity([self.scale['x'], self.scale['y']])
```
Comment (astrofrog) [inline sunpy/map/mapbase.py]: Why are these set to quantities? This has no impact, right?

File: sunpy/map/mapbase.py (hunk 1 above)
Comment (astrofrog) [inline sunpy/map/mapbase.py]: A future-proof way of writing this would be to do:

```
x = u.Quantity(x, unit=self.wcs.wcs.cunit[0])
```

Comment (Cadair) [inline sunpy/map/mapbase.py]: I tried that, but it was still returning arcsec.

File: sunpy/map/mapbase.py (hunk 2 above)
Comment (Cadair) [inline sunpy/map/mapbase.py]: No, but because both inputs are Quantities, making a Quantity array works better than a direct numpy array.

**Comment by Cadair** [comment]
supersedes and closes #1128.

File: sunpy/map/mapbase.py (hunk 2 above)
Comment (ehsteve) [inline sunpy/map/mapbase.py]: this will always be confusing, maybe a comment explaining this would be helpful

**Comment by Cadair** [comment]
@wafels now supporting datetime.

**Comment by wafels** [comment]
thanks!

**Comment by Cadair** [comment]
Just this left people. Then 0.6 can happen.

File: sunpy/map/mapbase.py [hunk 3]
```diff
@@ -772,20 +852,39 @@ def submap(self, range_a, range_b, units="data"):
...
-        >>> aia.submap([0,5],[0,5], units='pixels')
+        >>> aia.submap([0,5]*u.pixel, [0,5]*u.pixel)
         AIAMap([[ 0.3125, -0.0625, -0.125 ,  0.    , -0.375 ],
         [ 1.    ,  0.1875, -0.8125,  0.125 ,  0.3125],
         [-1.1875,  0.375 , -0.5   ,  0.25  , -0.4375],
         [-0.6875, -0.3125,  0.8125,  0.0625,  0.1875],
         [-0.875 ,  0.25  ,  0.1875,  0.    , -0.6875]])
         """
+        
+        # Do manual Quantity input validation
```
Comment (wafels) [inline sunpy/map/mapbase.py]: Add a note as to why no @astropy.units.quantity_input decorator?

File: sunpy/map/tests/test_mapbase.py [hunk 4]
```diff
@@ -230,30 +231,31 @@ def test_swap_cd():
...
+#    np.testing.assert_allclose(generic_map.data_to_pixel(generic_map.center['x'], generic_map.center['y']),
+#                              ((generic_map.meta['naxis1'] - 1) / 2., (generic_map.meta['naxis2'] - 1) / 2.))
 
     # Check conversion of map edges
     # Note: data coords are at pixel centers, so edges are 0.5 pixels wider
-    assert generic_map.data_to_pixel(generic_map.xrange[0], 'x') == 0. - 0.5
-    assert generic_map.data_to_pixel(generic_map.yrange[0], 'y') == 0. - 0.5
-    assert generic_map.data_to_pixel(generic_map.xrange[1], 'x') == (generic_map.meta['naxis1'] - 1) + 0.5
-    assert generic_map.data_to_pixel(generic_map.yrange[1], 'y') == (generic_map.meta['naxis2'] - 1) + 0.5
+#    assert generic_map.data_to_pixel(generic_map.xrange[0]) == 0. - 0.5
```
Comment (wafels) [inline sunpy/map/tests/test_mapbase.py]: why are these commented out and not deleted?

File: sunpy/map/tests/test_mapbase.py [hunk 5]
```diff
@@ -230,30 +231,31 @@ def test_swap_cd():
...
-    assert generic_map.data_to_pixel(generic_map.meta['crval1'], 'x') == generic_map.meta['crpix1'] - 1
-    assert generic_map.data_to_pixel(generic_map.meta['crval2'], 'y') == generic_map.meta['crpix2'] - 1
+    np.testing.assert_allclose(generic_map.data_to_pixel(generic_map.meta['crval1']*u.arcsec,
+                                                          generic_map.meta['crval2']*u.arcsec),
+                               np.array([generic_map.meta['crpix1'] - 1, generic_map.meta['crpix2'] - 1]))
 
     # Check conversion of map center
-    assert generic_map.data_to_pixel(generic_map.center['x'], 'x') == (generic_map.meta['naxis1'] - 1) / 2.
-    assert generic_map.data_to_pixel(generic_map.center['y'], 'y') == (generic_map.meta['naxis2'] - 1) / 2.
+#    np.testing.assert_allclose(generic_map.data_to_pixel(generic_map.center['x'], generic_map.center['y']),
```
Comment (wafels) [inline sunpy/map/tests/test_mapbase.py]: Why not delete the commented out code?

**Comment by ehsteve** [comment]
Are we going to add units to the data array? For AIA it is technically data numbers.

**Comment by Cadair** [comment]
You mean the primary array? That is encoded in the NDData.unit attribute. We could do it, is there a standard fits keyword for it?

**Comment by larrymanley** [comment]
FITS keyword PIXLUNIT is a string valued "DN" for AIA.  Could have other values post-processing, e.g. "FLUX" after normalizing for exposure time.

**Comment by dpshelio** [comment]
@larrymanley you were faster than me!!! I was just checking it from the [fits header documentaion](https://www.lmsal.com/sdodocs/doc?cmd=dcur&proj_num=SDOD0019&file_type=pdf). We may want to link to that document on the code...

**Comment by ehsteve** [comment]
@dpshelio it is already available in the header under the keyword 'keywddoc'

**Comment by dpshelio** [comment]
@ehsteve yeah, and it was there where I found it... but while reading the code you don't know why or where that is ;)

**Comment by Cadair** [comment]
Can we reliably convert from the header to astropy units?

**Comment by ehsteve** [comment]
@Cadair we should, at the very least, be able to do it on a case by case basis. I've started doing this for aia updating your pr.

**Comment by Cadair** [comment]
Final review on this one please.

**Comment by ayshih** [comment]
Fix wrapping of angular outputs from `Map.pixel_to_data()`

File: sunpy/map/mapbase.py [hunk 6]
```diff
@@ -154,32 +154,34 @@ def __repr__(self):
...
-Dimension:\t [{xdim:d}, {ydim:d}]
-[dx, dy] =\t [{dx:f}, {dy:f}]
+Dimension:\t {dim}
+scale:\t\t [{dx}, {dy}]
 
 """.format(dtype=self.__class__.__name__,
            obs=self.observatory, inst=self.instrument, det=self.detector,
            meas=self.measurement, date=self.date, dt=self.exposure_time,
-           xdim=self.data.shape[1], ydim=self.data.shape[0],
+           dim=self.shape,
```
Comment (ayshih) [inline sunpy/map/mapbase.py]: This change transposes the reported dimensions of the map, which seems unintended.

File: sunpy/map/mapbase.py [hunk 7]
```diff
@@ -458,41 +478,75 @@ def _validate(self):
...
+            Pixel coordinate on the CTYPE1 axis.
+
+        y : float
+            Pixel coordinate on the CTYPE2 axis.
+        """
+        x, y = self.wcs.wcs_world2pix(u.Quantity([x, y])[None, :].to(u.deg).value, origin)[0]
 
         return x, y
 
+    def pixel_to_data(self, x, y, origin=0):
```
Comment (ayshih) [inline sunpy/map/mapbase.py]: Should this take pixel Quantity inputs?

**Comment by Cadair** [comment]
The wrapping is now fixed and map.shape is no more.

File: sunpy/map/mapbase.py [hunk 8]
```diff
@@ -557,10 +613,8 @@ def resample(self, dimensions, method='linear'):
                                     method, center=True)
         new_data = new_data.T
 
-        # Note that 'x' and 'y' correspond to 1 and 0 in self.shape,
-        # respectively
-        scale_factor_x = (float(self.shape[1]) / dimensions[0])
-        scale_factor_y = (float(self.shape[0]) / dimensions[1])
+        scale_factor_x = float(self.dimensions[0] / dimensions[0])
+        scale_factor_y = float(self.dimensions[0] / dimensions[1])
```
Comment (ayshih) [inline sunpy/map/mapbase.py]: Typo, should be `scale_factor_y = float(self.dimensions[1] / dimensions[1])`

**Comment by ayshih** [comment]
The call to `self.rsun_arcseconds` in `Map.draw_limb()` needs to be fixed

File: sunpy/map/mapbase.py [hunk 9]
```diff
@@ -650,10 +704,47 @@ def rotate(self, angle=None, rmatrix=None, order=4, scale=1.0,
...
 
         # Interpolation parameter sanity
         if order not in range(6):
             raise ValueError("Order must be between 0 and 5")
+        
+        # is None to prevent 0 arrays evaling to False
+        if rotation_center is None:
+            rotation_center = u.Quantity([0*self.units['x'], 0*self.units['y']])
+        else:
+            if not isinstance(rotation_center, u.Quantity) and not hasattr(rotation_center, "unit"):
```
Comment (ayshih) [inline sunpy/map/mapbase.py]: Are you intending to disallow a list/tuple of Quantity objects?

File: sunpy/map/mapbase.py [hunk 10]
```diff
@@ -650,10 +704,47 @@ def rotate(self, angle=None, rmatrix=None, order=4, scale=1.0,
...
+                raise TypeError("Argument rotation_center to function rotate has"
+                                " no unit attribute. You may want to pass in an" 
+                                "astropy Quantity instead." )
```
Comment (ayshih) [inline sunpy/map/mapbase.py]: Needs a space before "astropy" so it doesn't render as "anastropy".

File: sunpy/map/mapbase.py (hunk 9 above)
Comment (Cadair) [inline sunpy/map/mapbase.py]: Not explicitly, but I am not sure that is a pattern we should support everywhere. That would be like supporting a tuple of numpy arrays, which would be a little odd.

Comment (ayshih) [inline sunpy/map/mapbase.py]: Putting aside the question of whether this pattern should be supported generally, in this particular instance, it's "natural" for the input to be a list/tuple of a longitude and a latitude, for example the output from `pixel_to_data()`.  The user could put the tuple through the `u.Quantity()` constructor to get something that will be accepted, but is the user supposed to know that?

**Comment by ayshih** [comment]
:+1:

I'd like further discussion on how we handle lists/tuples of Quantity objects, but that's not a blocking concern for this PR.

File: sunpy/map/mapbase.py (hunk 9 above)
Comment (Cadair) [inline sunpy/map/mapbase.py]: I see your point, but it's actually a really ugly thing to check for.

Currently, the way u.quantity_input works is that it checks for something that looks like a Quantity and that is convertible to the correct units.

I am not sure exactly what we should catch, apart from feeding the whole thing through the Quantity initializer and therefore making the allowed inputs quite different from other places in sunpy.

**Comment by Cadair** [comment]
@ehsteve @dpshelio @wafels Review when you get a min please?!

File: sunpy/map/mapbase.py [hunk 11]
```diff
@@ -61,7 +62,7 @@ class GenericMap(NDData):
     cmap : matplotlib.colors.Colormap
         A color map used for plotting with matplotlib.
     mpl_color_normalizer : matplotlib.colors.Normalize
-        A matplotlib normalizer used to scale the image plot.
+        A matplotlib normalizer used to the image plot.
```
Comment (wafels) [inline sunpy/map/mapbase.py]: This sentence does not make sense.

**Comment by Cadair** [comment]
@wafels fixed.

**Comment by wafels** [comment]
The superpixel method was not working.  I've made a PR to your branch to fix it.

**Comment by Cadair** [comment]
@wafels merged

**Comment by wafels** [comment]
+1 on your PR

**Comment by Cadair** [comment]
ok, so I have done some more testing of rotate, and this seems to break when you take a rotated image, submap it and then rotate the submap. While the submapping appears to be doing the right thing, draw grid will not plot on the final image properly:

```python
aiamap = sunpy.map.Map('/home/stuart/Dropbox/Stu and Lenny/aiacalibim5.fits')

sub1 = aiamap.submap([1000,2500]*u.pixel, [500,1500]*u.pixel)
sub1_rot = sub1.rotate(order=3)
# Plot original image against the rotated image
fig = plt.figure(figsize=(12, 9))
fig.add_subplot(1, 2, 1)
sub1.plot()
#sub1.draw_grid()
plt.title('Original image')
fig.add_subplot(1, 2, 2)
sub1_rot.plot()
sub1_rot.draw_grid()
plt.title('Image rotated 90 deg')
plt.show()
```

index

**Comment by Cadair** [comment]
```python
aiamap = sunpy.map.Map('/home/stuart/Dropbox/Stu and Lenny/aiacalibim5.fits')

sub1 = aiamap.submap([-1000,1000]*u.arcsec, [500,1000]*u.arcsec)
sub1_rot = sub1.rotate(order=3)
# Plot original image against the rotated image
fig = plt.figure(figsize=(12, 9))
fig.add_subplot(1, 2, 1)
sub1.plot()
sub1.draw_grid()
plt.title('Original image')
fig.add_subplot(1, 2, 2)
sub1_rot.plot()
sub1_rot.draw_grid()
plt.title('Image rotated 90 deg')
plt.show()
```
index

**Comment by ayshih** [comment]
There's a possible bug with `rotate()`?  Tell me more! =P

**Comment by Cadair** [comment]
@ayshih WOOOP!

I am thinking it is a bug in the way crpix is being calculated in submap. I am not sure though.

**Comment by ayshih** [comment]
Hmm.  Because `submap()` calls a now-orientation-aware `data_to_pixel()', calling `submap()` on a non-oriented map acts in weird ways when specifying physical coordinates.  For example, in the "original" images you show above, you can see that the region you may think you're asking for is not the region you get.

File: sunpy/map/mapbase.py [hunk 12]
```diff
@@ -797,29 +909,37 @@ def submap(self, range_a, range_b, units="data"):
...
+            y_pixels = range_b.value
         else:
             raise ValueError(
                 "Invalid unit. Must be one of 'data' or 'pixels'")
 
+        x_pixels = np.array(x_pixels)
+        y_pixels = np.array(y_pixels)
+        # Clip pixel values to max of array, prevents negative
+        # indexing
+        x_pixels[np.greater(x_pixels, self.data.shape[1])] = self.data.shape[1]
```
Comment (ayshih) [inline sunpy/map/mapbase.py]: Missing the corresponding clip protection for negative X pixels.

File: sunpy/map/mapbase.py [hunk 13]
```diff
@@ -1018,7 +1138,7 @@ def peek(self, draw_limb=False, draw_grid=False, gamma=None,
         ----------
         draw_limb : bool
             Whether the solar limb should be plotted.
-        draw_grid : bool or number
+         : bool or number
```
Comment (ayshih) [inline sunpy/map/mapbase.py]: Unintended deletion?

**Comment by Cadair** [comment]
@ayshih yes, it is correct, as long as you think of it as specifying two points in pixel space, there is nothing we can do about that.

The problem is the behaviour of the post-submap rotate() call. I have done a little testing with wcsaxes and it seems that the coords are correct after the submap:

index1

index2

but not after the rotate:

index3

**Comment by ayshih** [comment]
> @ayshih yes, it is correct, as long as you think of it as specifying two points in pixel space, there is nothing we can do about that.

I feel this is a usability problem.  Without wcsaxes, the user will be looking at the displayed coordinates in the rotated space.  If the user asks for a submap based on those displayed coordinates (rather than the underlying physical/WCS coordinates that he/she can't see), he/she won't get back the region he/she wants, but rather a different part of the Sun.  If we keep it this way, I think we need to at least throw a warning lest a user think that SunPy is randomly garbling his/her inputs.

**Comment by Cadair** [comment]
@ayshih that's a problem with plot, plot is broken unless the array is not rotated.

**Comment by Cadair** [comment]
so I added in wcsaxes support, it doesn't rely on the coordinates package it just uses the native wcs object. this means things are plotted correctly now.

**Comment by ayshih** [comment]
Um, are you sure that the reference coordinate is intended to have special meaning as the center of the transform?  If so, I shouldn't rewrite it the way I do in `rotate()`.

**Comment by Cadair** [comment]
well this is what @astrofrog told me. 

I mainly removed it because it broke the wcs information if you specified it, which is kind of the point of map.

**Comment by Cadair** [comment]
to expand on my wcsaxes comment:


I have pulled in what I was doing in #1237 which dealt with making all the wcsaxes stuff optional. then what I have done now is just use our draw_grid and draw_limb to plot overlays rather than use wcsaxes's coordinate support. this means that I plot, draw_limb, draw_grid now all work with and without wcsaxes. 

I need to test peek().

**Comment by ayshih** [comment]
I just fixed `rotate()` and put up a PR to your repo.

**Comment by ayshih** [comment]
Also, we have a problem with how we're handling the reference pixel/coordinate.  CRPIX1/2 in the FITS header starts counting at 1, but of course, Python starts at 0, as does it seems Astropy WCS.  That puts us in the weird situation where calling `data_to_pixel()` on the reference coordinate does not return the reference pixel, and the same problem with `pixel_to_data()`.

My suggestion is to have our `meta` keys match the FITS header (start counting at 1), but have the reference pixel property of Map automatically subtract 1 so that its count starts at 0.  This isolates the FITS convention, and makes everything outside of `meta` correct and consistent.

**Comment by Cadair** [comment]
ah interesting. that behavior in astropy wcs is controlled by the origin parameter which I just set to 0, thinking users would pass in numpy style coords.

I actually removed the origin kwarg from our api, but I will put it back so that we can use it for internal calls. Can you give a specific example to test with?

**Comment by Cadair** [comment]
@ayshih This is the trick:

```python
aiamap.data_to_pixel(*aiamap.reference_coordinate.values(), origin=1)
Out[9]: (<Quantity 512.5 pix>, <Quantity 512.4999999999948 pix>)

aiamap.reference_pixel
Out[10]: {'x': <Quantity 512.5 pix>, 'y': <Quantity 512.5 pix>}
```

**Comment by ayshih** [comment]
`origin` seems a bit weird, but I can roll with it.

I put up another PR to your repo

**Comment by Cadair** [comment]
@ayshih it's the integer value of the origin of the array, but more over it's the same as the astropy API!

**Comment by ayshih** [comment]
Oh, I meant that it's a bit weird in that it may confuse users.  I don't have a problem with the API or name.

**Comment by Cadair** [comment]
oh, yeah it is weird...

**Comment by ayshih** [comment]
Incidentally, I have a code concern given some of the code in the PR.  You've used the `.values()` method to pull out the values in a dictionary (relevant to our Map properties, e.g., `reference_coordinate`), and I've followed suit, but I'm worried that Python doesn't guarantee the order of the output.  Thoughts?

In fact, is there any benefit to having those Map properties as dictionaries in the first place – they all have keys that are *always* "x" and "y" – rather than tuples/lists?

**Comment by Cadair** [comment]
@ayshih yeah that's a good point actually.

maybe NamedTuples would be better?

**Comment by Cadair** [comment]
figure_1

**Comment by ayshih** [comment]
:+1:

File: doc/source/guide/plotting.rst [hunk 14]
```diff
@@ -160,15 +160,37 @@ using the show() command. Here is another example ::
 
     plt.show()
 
+By default `~sunpy.map.Map` uses the `wcsaxes <http://wcsaxes.readthedocs.org/>`_
```
Comment (dpshelio) [inline doc/source/guide/plotting.rst]: is `wcsaxes` a dependency now? should not that be added also in `setup.py`

**Pull request merged**